ops/_main.py

```diff
@@ -515,19 +499,48 @@ def _emit(self):
             self.framework.reemit()
 
         # Emit the Juju event.
-        _emit_charm_event(self.charm, self.dispatcher.event_name, self._juju_context)
+        self._emit_charm_event(self.dispatcher.event_name)
         # Emit collect-status events.
         ops.charm._evaluate_status(self.charm)
 
+    def _get_event_to_emit(self, event_name: str) -> Optional[ops.framework.BoundEvent]:
+        try:
+            return getattr(self.charm.on, event_name)
+        except AttributeError:
+            logger.debug('Event %s not defined for %s.', event_name, self.charm)
+        # If the event is not defined, return None.
+
+    def _emit_charm_event(self, event_name: str):
+        """Emits a charm event based on a Juju event name.
+
+        Args:
+            charm: A charm instance to emit an event from.
+            event_name: A Juju event name to emit on a charm.
+            juju_context: An instance of the _JujuContext class.
+        """
+        event_to_emit = self._get_event_to_emit(event_name)
+
+        # If the event is not supported by the charm implementation, do
+        # not error out or try to emit it. This is to support rollbacks.
+        if event_to_emit is not None:
+            args, kwargs = _get_event_args(self.charm, event_to_emit, self._juju_context)
+            logger.debug('Emitting Juju event %s.', event_name)
+            event_to_emit.emit(*args, **kwargs)
+
     def _commit(self):
         """Commit the framework and gracefully teardown."""
         self.framework.commit()
 
+    def _close(self):
+        """Perform any necessary cleanup before the framework is closed."""
+        # Provided for child classes - nothing needs to be done in the base.
```

Comment:
Comment could go in docstring?

Reply:
I feel like it's not part of the documentation of the method, it's commenting on the (lack of) implementation, so a comment is more appropriate.